### ISSUE TITLE
Beans that are AutoCloseable by instance are not closed.

### ISSUE THREAD

**Issue by agentgt** [avaje/avaje-inject#290]
For various reasons we often have a factory that will make a bean whose primary interface we want to expose but the concrete class also implements `AutoCloseable`. It is best probably to show this by example:

`HikariDataSource` is a special `DataSource` in that it also implements `AutoCloseable`. This is natural because its pool needs to be shutdown. 

However our factory method `@Bean` returns `DataSource` and not the specific `HikariDataSource` so I assume the generated code does not know this DataSource needs to be closed.

We do this in several other services as well.

It seems like the lifecycle handling of `BeanScope` should `instanceof` all instantiated beans to see if they are  closeable and close them. That or perhaps some configuration.

I can't recall if Spring does this but I think other DIs do (our in house manual DI does).

**Comment by SentryMan** [comment]
Currently, the closable detection works at compile time by reading the actual return type of the factory method/singleton. 

> However our factory method @Bean returns DataSource and not the specific HikariDataSource

Might I ask why? If you return the specific HikariDataSource it's still registered as a DataSource in the scope and will have the autoclose functionality. Your other beans can still wire using the primary DataSource interface. 

Even so, I'll see if I can make it work at runtime.

**Comment by rbygrave** [comment]
So with the PR just merged we can do:

```java
@Factory
class MyFactory  {

  @Bean(autoCloseable=true)
  DataSource provideDataSource(...) {
    return ...
  }

```
And then the generated code will add a `builder.addAutoClosable($bean)` after the bean is registered like:
```java
      var $bean = builder.register(bean);
      builder.addAutoClosable($bean);
```
... and if the instance is AutoCloseable then that will register with PreDestroy.


**Background:**
The thinking here is that this issue of need to do `bean instanceOf AutoCloseable` is restricted to beans that are created via `@Factory` `@Bean`.  The `@Bean` already has the `initMethod` and `destroyMethod` attributes (which are also common with Spring DI and Micronaut DI) to register lifecycle methods.  

So the thinking here is that the new `autoCloseable` attribute on `@Bean` is reasonably in line with the different but related `destroyMethod` attribtue - they are the 2 ways to tell the DI that we have a PreDestroy method that we want invoked when the DI context is closed.

**Comment by agentgt** [comment]
> The thinking here is that this issue of need to do bean instanceOf AutoCloseable is restricted to beans that are created via @Factory @Bean. 

I agree and am massively pleased with the changes.

**Comment by rbygrave** [comment]
FYI: This change is in version **_8.13-RC4_** which is on it's way to maven central now.

**Comment by rbygrave** [comment]
Closing fixed in 8.13
